Create Main Component

TYMG opened the pull request:
* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)

This PR adds a header, footer and main div to the app-component

* **What is the current behavior?** (You can also link to an open issue here)

Currently, the behavior of the app is incorrect since the content was old code.

* **What is the new behavior (if this is a feature change)?**

There is a responsive header and footer. And the main div has a simple p tag.

* **Other information**:
I wrote a couple unit tests. I wasn't sure how in depth the tests were supposed to be for something so simple. Using query selectors, I tested for HTML Elements. And there is a test to create the AppComponent as well.

TYMG committed:
Pushing LDD Template

TYMG committed:
Updated the LDD to add more detail. Will review.

TYMG committed:
Added Typings Cleaned Code;

TYMG committed:
3/3 Unit Test Working. Style/Conent Header Footer

TYMG committed:
Completed Building Base App

chriscarmel commented on client/karma.conf.js:
```diff
@@ -4,17 +4,28 @@
     client:{
       clearContext: false // leave Jasmine Spec Runner output visible in browser
-    },
+    },     
```
Looks like you've added some rogue whitespace at the end of this line. You can prevent this by changing this setting in your [VSCode preferences](https://code.visualstudio.com/docs/getstarted/settings):
```
// When enabled, will trim trailing whitespace when saving a file.
"files.trimTrailingWhitespace": true,
```

chriscarmel commented on client/src/app/app.component.scss:
```diff
@@ -0,0 +1,37 @@
+    width: 100%;
+    /* Set the fixed height of the footer here */
+    height: 60px;
+    line-height: 60px; /* Vertically center the text there */
```
This looked funny to me, but I read up about it on [this minimalist page](http://www.vanseodesign.com/blog/demo/vertical-centering/line-height.php) and I understand now that it pads the container above and below the content. The one caveat was that it must be only a single, unbroken line, so as long as we're diligent about that, we'll be good.

chriscarmel commented on client/src/app/app.component.scss:
```diff
@@ -0,0 +1,37 @@
+body {
+    /* Margin bottom by footer height */
+    margin-bottom: 60px;
+  }
```
Looks like you've got some mismatched indentation in this file. VSCode will take care of this for you by auto-formatting on save if you change this setting in your [VSCode Preferences](https://code.visualstudio.com/docs/getstarted/settings):
```
 // Format a file on save. A formatter must be available, the file must not be auto-saved, and editor must not be shutting down.
  "editor.formatOnSave": false,
```

chriscarmel commented on client/src/app/app.component.html:
```diff
@@ -1,11 +1,19 @@
+            <a class="flex-sm-fill text-sm-center nav-link disabled" href="#">Disabled</a>
+          </nav>
+      </header>
+  <div class="app__main container-fluid h-75">
```
The [Bootstrap class](https://v4-alpha.getbootstrap.com/utilities/sizing/) `h-75` here applies a rule stating the element to have a height of 75% of that of the parent container, thus splitting the header and footer sibling tags into a height of 12.5% apiece. In the design document for this effort, the requirements specify that the header and footer get a height of 10% apiece. What gives?

chriscarmel commented on client/src/app/app.component.scss:
```diff
@@ -0,0 +1,37 @@
+// Standard Styles
+body {
+    /* Margin bottom by footer height */
+    margin-bottom: 60px;
```
The comment here seems to convey that this offset is being applied to account for the `footer` occupying `60px` at the bottom of the screen. The `footer` is a child of the `main` element, which has the `app__main` class, so why not define this rule in the `app__main` class?

chriscarmel commented on client/src/app/app.component.scss:
```diff
@@ -0,0 +1,37 @@
+    bottom: 0;
+    width: 100%;
+    /* Set the fixed height of the footer here */
+    height: 60px;
```
In the design document, the `footer` element height property is defined to be `10%` of the page height. Why have you defined it here with a hard-coded height of `60px`?

chriscarmel commented on client/src/app/app.component.scss:
```diff
@@ -0,0 +1,37 @@
+    /* Margin bottom by footer height */
+    margin-bottom: 60px;
+  }
+  .footer {
```
You've defined the `footer` class, but you haven't defined the `header` class as mentioned in the design document. Where is it?

chriscarmel commented on client/src/app/app.component.html:
```diff
@@ -1,47 +1,19 @@
-    </form>
+<main id="app__main" class="h-100">
+  <header class="header">
+        <nav class="nav nav-pills flex-column flex-sm-row">
```
I'm torn between marking or not marking this as work that doesn't belong in this branch because it wasn't mentioned in the design document, only this:

> Add a Static header and footer with a main div in between

And also because this line was mentioned:

> The App Component will bare as possible

However, the first quote just says a static header is present, so it could be interpreted that this is part of the static header that was implied, and the second quote is only defined under the controller section, so the "bare as possible" intent doesn't necessary refer to the template, so can't fault you here.

Not sure whether you want to take action regarding this or not, but in the future we can try to emphasize separation of concerns and managing scope creep or be more clear about intention with a little more description in the design document. What are your thoughts?

TYMG replied:
Let's discuss this in person, I think this is an issue with wording and interpretation.

chriscarmel commented:
All of the remarks I left on this PR so far have been [comments](https://help.github.com/articles/commenting-on-a-pull-request/) on lines of code, but I just learned a little more about [starting a review](https://help.github.com/articles/reviewing-proposed-changes-in-a-pull-request/) that begins an exchange with some sort of resolution functionality. I'm gonna start using that on the next PR or round of review on this one.

TYMG commented:
Chris, I added the VS Code settings, so hopefully, the rogue white space issue will be fixed going forward.

TYMG committed:
Change Code Review Comments

TYMG commented:
Chris, I updated the code base to better reflect the DD. Apologies, I used [the bootstrap sticky footer example](http://v4-alpha.getbootstrap.com/examples/sticky-footer-navbar/ ) for the styling. I changed the app.component.scss file so now the header and footer should be 10%. Let me know what you think. And the help is always appreciated man.

TYMG committed:
Code Review; Sass Auto formatter

TYMG committed:
Body Change Class Name
